### ISSUE TITLE
Requirements installation needs refinement

### ISSUE THREAD

**Issue by ChanceNCounter** [OpenVoiceOS/ovos_skill_manager#44]
* Should check whether deps are already present
* Should keep track during installation of what it's already seen in other deps.

Example of status quo, truncated and commented (long):

```
ovos_skills_manager.skill_entry:install:273 - INFO - Installing skill: https://github.com/JarbasSkills/skill-old-world-radio fr
om branch: v0.2
ovos_skills_manager.skill_entry:install:273 - INFO - Installing skill: https://github.com/JarbasSkills/skill-better-playback-co
ntrol from branch: v0.3a1
ovos_skills_manager.skill_entry:install:290 - INFO - Running pip install
ovos_skills_manager.requirements:pip_install:43 - INFO - (pip) Installing ovos_utils>=0.0.8 # system version is 0.0.9a2
ovos_skills_manager.requirements:pip_install:43 - INFO - (pip) Installing youtube-dl
ovos_skills_manager.requirements:pip_install:43 - INFO - (pip) Installing ovos_workshop>=0.0.1a3
ovos_skills_manager.requirements:pip_install:43 - INFO - (pip) Installing pafy
ovos_skills_manager.skill_entry:install:293 - INFO - Downloading https://github.com/JarbasSkills/skill-better-playback-control
ovos_skills_manager.github:get_requirements_json:209 - ERROR - Error reading from manifest!
ovos_skills_manager.github:get_requirements_json:209 - ERROR - Error reading from manifest!
ovos_skills_manager.skill_entry:install:290 - INFO - Running pip install
ovos_skills_manager.requirements:pip_install:43 - INFO - (pip) Installing ovos_utils>=0.0.8a1 # we just had this conversation
ovos_skills_manager.requirements:pip_install:43 - INFO - (pip) Installing youtube-dl # also this one
ovos_skills_manager.requirements:pip_install:43 - INFO - (pip) Installing pafy # and this one
ovos_skills_manager.skill_entry:install:293 - INFO - Downloading https://github.com/JarbasSkills/skill-old-world-radio
```

Programmatically, the current situation makes perfect sense, but it's a nightmare on slower devices. The console output is taunting me!

**Comment by NeonDaniel** [comment]
I haven't looked into this much; does the dependency installation include version conflict resolution? I'd think the best way would be to compile all skill dependencies and install in one shot if installing multiple skills (would be a re-structuring of installation though)

**Comment by JarbasAl** [comment]
I fully agree, i would also like to add support for conflict resolution, track requirements for each installed skill externally and whenever a package would be updated inform the use, eg:

- skill A installs pkgA==0.0.1
- skill B installs pkgB==0.1.2 and pkgA==0.0.2
- pip doesnt detect any error, but skill A now broke

we can use semantic versioning to be somewhat smart about this or just ask the user every time, depending on the specific skills it might be possible o find a compatible version or not

**Comment by NeonDaniel** [comment]
> * skill A installs pkgA==0.0.1
> * skill B installs pkgB==0.1.2 and pkgA==0.2.2

Doesn't pip log a warning in that case now, or is that only for dependencies relating to installed packages?

**Comment by JarbasAl** [comment]
pip will notice this stuff in a single requirements.txt file, but these are install steps in different points in time, skills are not individually tracked by pip, they are not python modules, for all that pip is concerned he sees the previous packages installed but nothing is using them

**Comment by NeonDaniel** [comment]
Looking into this, is there a reason we are enforcing order in dependency installation? Seems like just passing the full list would be marginally faster (merging multiple skills dependencies into one list of packages would be even better). Existing osm-installed skills could get constraint files generated to prevent breaking versions (would need exception handling to notify user what failed.

**Comment by ChanceNCounter** [comment]
:+1: I think there are a few ways in which OSM needs to behave more like a package manager, especially since OVOS devices themselves are primarily customized through skills and plugins.

**Comment by JarbasAl** [comment]
> Looking into this, is there a reason we are enforcing order in dependency installation? 

the only reasoning was that if a requirement skill failed to install, then the actual skill install wouldnt even be attempted

**Comment by NeonDaniel** [comment]
> > Looking into this, is there a reason we are enforcing order in dependency installation?
> 
> the only reasoning was that if a requirement skill failed to install, then the actual skill install wouldnt even be attempted

But for pip dependencies, we should be able to just throw everything at pip at once (per-skill), right?

**Comment by ChanceNCounter** [comment]
I think we can do better than per-skill, if we do it in passes.
